File: src/main/java/dev/klomptech/jbark/parser/controlflow/ForEachParser.java
package dev.klomptech.jbark.parser.controlflow;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.Parser;
import java.util.ArrayList;
import java.util.List;

public class ForEachParser {

  private final Parser parser;

  private record ForEachLineMarkers(int thenOffset) {}

  public ForEachParser(final Parser parser) {
    this.parser = parser;
  }

  // Example: "for each treat from her cookie jar then"
  public AstNode parseForEachLine() throws BarkError {
    int line = parser.peek().line();
    if (!parser.currentLineStartsWithAny(Keywords.FOR_WORDS)) {
      return null;
    }
    if (findWordOffset(Keywords.EACH_WORDS) < 0) {
      throw new BarkError(
          line, "For loops start with for each, like 'for each bone from cookie jar'");
    }
    int eachAt = findWordOffset(Keywords.EACH_WORDS);
    // the variable name after "each"
    String variable = findLoopVariableAfter(eachAt);
    if (variable == null) {
      throw new BarkError(line, "For each what? Name the loop bone or treat.");
    }
    int fromAt = findFromOrInAfter(eachAt + 1);
    if (fromAt < 0) {
      throw new BarkError(line, "For each needs from or in, like 'for each bone from cookie jar'");
    }
    // Which stash to walk through
    String stash =
        ConfigLoader.resolveCollectionFromTokens(
            parser.allTokens(), parser.cursorIndex() + fromAt + 1, parser.lineEndIndex());
    if (stash == null || !ConfigLoader.isStash(stash)) {
      throw new BarkError(line, "After from, name a stash the dogs can dig through");
    }
    ConfigLoader.requireStash(stash, line);
    ForEachLineMarkers markers = findForEachLineMarkers();
    if (markers.thenOffset() < 0) {
      throw new BarkError(line, "The dogs heard the start of a command but no 'then'");
    }
    // Read the lines below "then" (or one print on the same line) until bury
    List<AstNode> steps = parseLoopBody(markers, line);
    if (!parser.currentLineIsBury()) {
      throw new BarkError(line, "The dog started something but forgot what he was doing");
    }
    parser.consumeBuryLine();
    return new AstNode.ForEach(variable, stash, steps, line);
  }

  // First identifier after "each" and before "from" or "in" (example: treat in "for each treat
  // from")
  private String findLoopVariableAfter(final int eachAt) {
    for (int offset = eachAt + 1; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.FROM_WORDS.contains(word) || Keywords.IN_WORDS.contains(word)) {
        return null;
      }
      if (Keywords.isThenWord(word)) {
        return null;
      }
      return word;
    }
    return null;
  }

  private int findFromOrInAfter(final int startOffset) {
    for (int offset = startOffset; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.FROM_WORDS.contains(word) || Keywords.IN_WORDS.contains(word)) {
        return offset;
      }
    }
    return -1;
  }

  private ForEachLineMarkers findForEachLineMarkers() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER) && Keywords.isThenWord(parser.normalise(token.value()))) {
        return new ForEachLineMarkers(offset);
      }
    }
    return new ForEachLineMarkers(-1);
  }

  // Find one of the words in the list
  private int findWordOffset(final List<String> words) {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER) && words.contains(parser.normalise(token.value()))) {
        return offset;
      }
    }
    return -1;
  }

  // Build the list of steps inside the for each block (multiline lines below "then", or one print
  // on the same line)
  private List<AstNode> parseLoopBody(final ForEachLineMarkers markers, final int line)
      throws BarkError {
    // Multiline: "for each treat from her cookie jar then" and the body starts on the next line
    if (parser.isAtEndOrLineEndOffset(markers.thenOffset() + 1)) {
      parser.advanceBy(markers.thenOffset() + 1);
      parser.skipNewlines();
      return parseMultilineBody(line);
    }
    // One line: "for each treat from her cookie jar then she woofs x" on one line (print only)
    int tokenCount = parser.countTokensAhead(markers.thenOffset() + 1);
    parser.advanceBy(markers.thenOffset() + 1);
    return InlineBodyParser.parseThenSteps(
        parser, tokenCount, line, "Then what? The dogs are waiting for something to do.");
  }

  // Body lines below "for each-then" until a bury line is found
  private List<AstNode> parseMultilineBody(final int line) throws BarkError {
    List<AstNode> steps = new ArrayList<>();
    while (!parser.isAtEnd()) {
      parser.skipNewlines();
      if (parser.isAtEnd()) {
        break;
      }
      if (parser.currentLineIsBury()) {
        break;
      }
      parser.parseStatementInto(steps);
    }
    if (steps.isEmpty()) {
      throw new BarkError(line, "Then what? The dogs are waiting for something to do.");
    }
    return steps;
  }
}

File: pbark/pbark/parser/controlflow/foreach_parser.py
from __future__ import annotations

from dataclasses import dataclass

from pbark.config.config_loader import ConfigLoader
from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords
from pbark.parser.controlflow.inline_body_parser import InlineBodyParser


@dataclass
class _ForEachLineMarkers:
    then_offset: int


class ForEachParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_for_each_line(self):
        line = self._parser.peek().line
        if not self._parser.current_line_starts_with_any(Keywords.FOR_WORDS):
            return None
        if self._find_word_offset(Keywords.EACH_WORDS) < 0:
            raise BarkError(
                line,
                "For loops start with for each, like 'for each bone from cookie jar'",
            )
        each_at = self._find_word_offset(Keywords.EACH_WORDS)
        variable = self._find_loop_variable_after(each_at)
        if variable is None:
            raise BarkError(line, "For each what? Name the loop bone or treat.")
        from_at = self._find_from_or_in_after(each_at + 1)
        if from_at < 0:
            raise BarkError(line, "For each needs from or in, like 'for each bone from cookie jar'")
        stash = ConfigLoader.resolve_collection_from_tokens(
            self._parser.all_tokens(),
            self._parser.cursor_index() + from_at + 1,
            self._parser.line_end_index(),
        )
        if stash is None or not ConfigLoader.is_stash(stash):
            raise BarkError(line, "After from, name a stash the dogs can dig through")
        ConfigLoader.require_stash(stash, line)
        markers = self._find_for_each_line_markers()
        if markers.then_offset < 0:
            raise BarkError(line, "The dogs heard the start of a command but no 'then'")
        steps = self._parse_loop_body(markers, line)
        if not self._parser.current_line_is_bury():
            raise BarkError(line, "The dog started something but forgot what he was doing")
        self._parser.consume_bury_line()
        return AstNode.ForEach(variable, stash, steps, line)

    def _find_loop_variable_after(self, each_at: int) -> str | None:
        offset = each_at + 1
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                offset += 1
                continue
            word = self._parser.normalise(token.value)
            if word in Keywords.FROM_WORDS or word in Keywords.IN_WORDS:
                return None
            if Keywords.is_then_word(word):
                return None
            return word
        return None

    def _find_from_or_in_after(self, start_offset: int) -> int:
        offset = start_offset
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                offset += 1
                continue
            word = self._parser.normalise(token.value)
            if word in Keywords.FROM_WORDS or word in Keywords.IN_WORDS:
                return offset
            offset += 1
        return -1

    def _find_for_each_line_markers(self) -> _ForEachLineMarkers:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and Keywords.is_then_word(
                self._parser.normalise(token.value)
            ):
                return _ForEachLineMarkers(offset)
            offset += 1
        return _ForEachLineMarkers(-1)

    def _find_word_offset(self, words: list[str]) -> int:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in words:
                return offset
            offset += 1
        return -1

    def _parse_loop_body(self, markers: _ForEachLineMarkers, line: int):
        if self._parser.is_at_end_or_line_end_offset(markers.then_offset + 1):
            self._parser.advance_by(markers.then_offset + 1)
            self._parser.skip_newlines()
            return self._parse_multiline_body(line)
        token_count = self._parser.count_tokens_ahead(markers.then_offset + 1)
        self._parser.advance_by(markers.then_offset + 1)
        return InlineBodyParser.parse_then_steps(
            self._parser,
            token_count,
            line,
            "Then what? The dogs are waiting for something to do.",
        )

    def _parse_multiline_body(self, line: int):
        steps = []
        while not self._parser.is_at_end():
            self._parser.skip_newlines()
            if self._parser.is_at_end():
                break
            if self._parser.current_line_is_bury():
                break
            self._parser.parse_statement_into(steps)
        if not steps:
            raise BarkError(line, "Then what? The dogs are waiting for something to do.")
        return steps
